src/python/zquantum/core/circuits/_decomposition.py

```diff
@@ -0,0 +1,42 @@
+from typing import Iterable, TypeVar
+
+from typing_extensions import Protocol
+
+OperationType = TypeVar("OperationType")
+
+
+class DecompositionRule(Protocol[OperationType]):
+    def production(self, operation: OperationType) -> Iterable[OperationType]:
+        raise NotImplementedError()
+
+    def predicate(self, operation: OperationType) -> bool:
+        raise NotImplementedError()
+
+
+def decompose_operation(
+    operation: Iterable[OperationType],
+    decomposition_rules: Iterable[DecompositionRule[OperationType]],
+):
+    rules_iter = iter(decomposition_rules)
```

Comment:
I remember that forking iterators with [itertools.tee](https://docs.python.org/3/library/itertools.html#itertools.tee) is prone to performance problems if the forked iterators diverge much. I'm worried we might face a similar issue here when decomposing big circuits. WDYT?

Reply:
Switch to using eager slices in 374c308